### ISSUE TITLE
A command for releasing packages from a monorepo

### ISSUE THREAD

**Issue by zkochan** [pnpm/pnpm#2225]
Related #1098

The proposed solution for the publishing tool.

When the command runs (`pnpm version`), it finds all the modified packages since last publish. It then lists the modified packages in an editor, suggesting to select the next versions (similar to `git rebase -i`). For instance:

```text
foo 1.0.0 => minor
bar 2.0.0 => 2.1.0
```

After saving the file, pnpm will either automatically continue (if it is possible to implement) or a command like `pnpm version --continue` will have to be run. On the next stage, pnpm will propose to release every dependent of the packages selected in the previous stage. The second step repeats until there are dependents left.

**Comment by tunnckoCore** [comment]
I have packages around that that can fit perfectly here :wink: 
They are conventional-commits oriented, but which can sit by some flag, or not promote it as default?

- [detect-next-version](https://ghub.now.sh/detect-next-version)
- [git-commits-since](https://ghub.now.sh/git-commits-since)
- [parse-commit-message](https://github.com/tunnckoCore/monorepo/tree/master/packages/parse-commit-message)
- [recommended-bump](https://github.com/tunnckoCore/monorepo/tree/master/packages/recommended-bump) (used in detect-next-version)
_(moving them soon to the [tunnckoCore/opensource](https://github.com/tunnckoCore/opensource) monorepo, where I'll try latest pnpm again)_

But yea, proposed approach would be good too. But probably only as default behavior.

**Comment by zkochan** [comment]
Seems like Yarn v2 has a pretty good implementation https://github.com/yarnpkg/berry/pull/682

**Comment by tunnckoCore** [comment]
Hey @zkochan any ETA on that? :tada:

**Comment by zkochan** [comment]
I have found also this project: https://github.com/atlassian/changesets

So I though, I'd look into if we can make that one work with pnpm.

**Comment by tuan231195** [comment]
@zkochan: do you have any examples/documentation of how to use changesets with pnpm?

**Comment by zkochan** [comment]
To initialize

```
pnpm add -DW @changesets/cli
pnpx changeset init
```

To create changesets

```
pnpx changeset
# commit the changes
```

To release the changes

```
pnpx changeset version
pnpm install
pnpm publish -r
# commit the changes
```

**Comment by diminutivesloop** [comment]
@zkochan What's the reasoning behind using `pnpm publish -r` instead of `pnpx changeset publish`?

**Comment by ChristianUlbrich** [comment]
What we really need is a dependency aware publishing of packages. Currently using `workspace:*`uses the _latest_ published package; but because `pnpm publish -r` simply publishes all packages in a monorepo in (so it seems) _alphabetical_ way, this won't work as expected.

Package A depending on Package B@`workspace:*`, will get published _before_ Package B and thus it will reference the older version of Package B, instead of the soon to be published one. This results in a different behavior if Package A is now consumed from the registry, as opposed to the consumption in the monorepo.  

The solution would be, that Package B gets published first, because it is a dependency of Package A, then Package A can get published with the now already published version of Package B.

Basically this is the same behavior as `pnpm exec -r` (with _implicitly_ `--sort true`) already provides.

**Comment by zkochan** [comment]
`workspace:*` should be replaced with the version that is present in the monorepo.

> What's the reasoning behind using pnpm publish -r instead of pnpx changeset publish?

changeset publish publishes in random order. `pnpm publish -r` publishes dependencies then dependents.

**Comment by ChristianUlbrich** [comment]
You are right, my bad.

**Comment by ericraio** [comment]
bump

**Comment by zkochan** [comment]
I am not sure. I think I would prefer to collaborate with changesets. But having a better tool integrated to pnpm would give us advantage over npm CLI.

**Comment by DavidWells** [comment]
I'm also curious to see something like this happen. 

The changeset flow is ok for CI but is a little cumbersome otherwise (IMO). 

I'm stuck in a tween state of using lerna to publish and `pnpm` for everything else. It seems like `lerna` and `pnpm` don't seem to like to live in the same repos due to the package-lock conflicts. I'll try to [disable](https://github.com/lerna/lerna/issues/1235#issuecomment-422288361) them and see if that goes better 😃

**Comment by mikestopcontinues** [comment]
I would prefer `pnpm version` exist. Changeset requires opting into way too much.

**Comment by lkimrxnt** [comment]
I was surprised that `pnpm version` is not a thing. Was wondering if it's possible to implement the command.

**Comment by alshdavid** [comment]
Would be nice if I could run `pnpm publish --access public --filter "*"` from my workspace root and have pnpm publish all packages that are not marked at `private: true`.

**Comment by zkochan** [comment]
@alshdavid that command will work. `-r` should be used instead of `--filter "*"` and it will work. But you should update the versions.

**Comment by zkochan** [comment]
Now that we have added [catalogs](https://github.com/pnpm/pnpm/issues/7072) to pnpm, tracking changes in projects is a bit harder.

Before catalogs changesets could detect when a dependency gets updated by simply diffing the project's directory. But now the dependency will only change in `pnpm-workspace.yaml`, so a diff of the directory will not change. We'd have to implement our own diffing.

By the way, there is one other alternative for releasing package that wasn't mentioned in the thread: [Bit](https://bit.cloud/). In a Bit workspace Bit automatically detects any modifications to the packages (components) and releases the changed packages and their dependents.

**Comment by ghoullier** [comment]
We do use lerna-lite to handle version publishing in our monorepo and it's unfortunately not compatible with catalogs. It would be nice to have a pnpm command to replace catalog with the right version.

**Comment by zkochan** [comment]
> We do use lerna-lite to handle version publishing in our monorepo and it's unfortunately not compatible with catalogs. It would be nice to have a pnpm command to replace catalog with the right version.

I don't know if lerna-lite can just handle versioning but if it could bump the versions in the package.json files w/o publishing anything, then you could run `pnpm publish -r` and pnpm will release all packages that were bumped (and replace and workspace and catalog protocols).

**Comment by ghoullier** [comment]
`lerna-lite` rely on [npm-package-arg](https://github.com/npm/npm-package-arg) to create a dependency graph of the packages tree to compute new packages versions. This step fails because it cannot resolve dependencies with catalog protocol.
We cannot use `pnpm publish -r` because in our workflow we use conventionnal-commits to bump packages version before publish.

**Comment by haoqunjiang** [comment]
FYI lerna-lite has added partial `catalog:` support in its most recent release https://github.com/lerna-lite/lerna-lite/releases/tag/v3.11.0 so it's usable again.

**Comment by ghiscoding** [comment]
hi everyone, I just read this discussion and I'm the Lerna-Lite maintainer. 

So just in case you didn't know yet, I added full pnpm catalog support in [v3.12.0](https://github.com/lerna-lite/lerna-lite/releases/tag/v3.12.0) and to make sure all is good, I'm also using pnpm catalog even in Lerna-Lite itself (default catalog that is, but named catalog also works correctly, I'm also using catalogs in other couple open source projects). 

_Side note, the original Lerna doesn't yet support pnpm catalog, it's only supported by Lerna-Lite_

Cheers and thanks for pnpm which is an awesome project... which is even more awesome when used in combo with [Lerna-Lite](https://github.com/lerna-lite/lerna-lite/) 😉

**Comment by zkochan** [comment]
Changesets no longer satisfy our own needs in the pnpm/pnpm repository. Hence, I work on an RFC to integrate it into pnpm: https://github.com/pnpm/rfcs/pull/18

**Comment by zkochan** [comment]
Done: https://pnpm.io/versioning
